src/Fl_ICO_Image.cxx

```diff
@@ -0,0 +1,189 @@
+//
+// Fl_ICO_Image class for the Fast Light Tool Kit (FLTK).
+//
+// Copyright 2020 by Bill Spitzak and others.
+//
+// This library is free software. Distribution and use rights are outlined in
+// the file "COPYING" which should have been included with this file.  If this
+// file is missing or damaged, see the license at:
+//
+//     https://www.fltk.org/COPYING.php
+//
+// Please report all bugs and problems on the following page:
+//
+//     https://www.fltk.org/str.php
+//
+
+
+//
+// Include necessary header files...
+//
+
+#include <FL/Fl.H>
+#include "config.h"
+#include "Fl_Image_Reader.h"
+#include <FL/Fl_ICO_Image.H>
+#if defined(HAVE_LIBPNG) && defined(HAVE_LIBZ)
+#  include <FL/Fl_PNG_Image.H>
+#endif
+
+Fl_ICO_Image::Fl_ICO_Image(const char *filename, const unsigned char *data, int id)
+: Fl_BMP_Image(0,0),
+ idcount_(0),
+ icondirentry_(0)
+{
+  Fl_Image_Reader rdr;
+  int r;
+
+  w(0); h(0); d(0); ld(0);
+  alloc_array = 0;
+  array = 0;
+
+  if (data) {
+    r = rdr.open(filename, data);
+  } else {
+    r = rdr.open(filename);
+  }
+
+  if (r == -1) {
+    ld(ERR_FILE_ACCESS);
+  } else {
+    load_ico_(rdr, id);
+  }
+}
+
+/*
+ This method attempts to load the biggest image resource available
+ inside a .ICO file (Windows Icon format).
+ */
+void Fl_ICO_Image::load_ico_(Fl_Image_Reader &rdr, int id)
+{
+  int pickedID = -1;
+
+  // Check file header (ICONDIR, 6 bytes)
+
+  if (rdr.read_word() != 0 || rdr.read_word() != 1) {
+    Fl::error("Fl_ICO_Image: %s is not an ICO file.\n", rdr.name());
+    ld(ERR_FORMAT);
+    return;
+  }
+
+  idcount_ = rdr.read_word();
+
+  if (idcount() == 0) {
+    Fl::error("Fl_ICO_Image: %s - no image resources found\n", rdr.name());
+    ld(ERR_FORMAT);
+    return;
+  }
+
+
+  // read entries (ICONDIRENTRY, 16 bytes each)
+
+  icondirentry_ = new fl_ICONDIRENTRY[idcount()];
+
+  for (int i = 0; i < idcount(); ++i) {
+    icondirentry_[i].bWidth = (int)rdr.read_byte();
+    icondirentry_[i].bHeight = (int)rdr.read_byte();
+    icondirentry_[i].bColorCount = (int)rdr.read_byte();
+    icondirentry_[i].bReserved = (int)rdr.read_byte();
+    icondirentry_[i].wPlanes = (int)rdr.read_word();
+    icondirentry_[i].wBitCount = (int)rdr.read_word();
+    icondirentry_[i].dwBytesInRes = (int)rdr.read_dword();
+    icondirentry_[i].dwImageOffset = (int)rdr.read_dword();
+
+    if (icondirentry_[i].bWidth == 0) icondirentry_[i].bWidth = 256;
+    if (icondirentry_[i].bHeight == 0) icondirentry_[i].bHeight = 256;
+  }
+
+  if (id <= -2) return;
```

Comment:
Adding a note that this return here is done to only read the `icondirentry_` data without the actual image data would be useful.